presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/ReorderJoins.java

```diff
@@ -0,0 +1,283 @@
+            PlanNode right = getJoinSource(
+                    idAllocator,
+                    rightSources,
+                    pushDownResult.getRightPredicate(),
+                    joinSymbols.stream()
+                            .filter(symbol -> !leftSymbols.contains(symbol))
+                            .collect(toImmutableList()));
+
+            List<Expression> joinPredicates = extractConjuncts(pushDownResult.getJoinPredicate());
+            List<JoinNode.EquiJoinClause> joinConditions = joinPredicates.stream()
+                    .filter(JoinEnumerator::isJoinEqualityCondition)
+                    .map(predicate -> toEquiJoinClause((ComparisonExpression) predicate, leftSymbols))
+                    .collect(toImmutableList());
+            List<Expression> joinFilters = joinPredicates.stream()
+                    .filter(predicate -> !isJoinEqualityCondition(predicate))
+                    .collect(toImmutableList());
+
+            // sort output symbols so that the left input symbols are first
+            List<Symbol> inputSymbols = ImmutableList.<Symbol>builder()
+                    .addAll(left.getOutputSymbols())
+                    .addAll(right.getOutputSymbols())
+                    .build();
+            List<Symbol> outputSymbols = inputSymbols.stream()
+                    .filter(joinGraph.getOutputSymbols()::contains)
+                    .collect(toImmutableList());
+            return new JoinNode(
+                    idAllocator.getNextId(),
+                    INNER,
+                    left,
+                    right,
+                    flipJoinCriteriaIfNecessary(joinConditions, left),
+                    outputSymbols,
+                    joinFilters.isEmpty() ? Optional.empty() : Optional.of(ExpressionUtils.and(joinFilters)),
+                    Optional.empty(),
+                    Optional.empty(),
+                    Optional.empty());
+        }
+
+        private static List<JoinNode.EquiJoinClause> flipJoinCriteriaIfNecessary(List<JoinNode.EquiJoinClause> joinCriteria, PlanNode left)
+        {
+            return joinCriteria
+                    .stream()
+                    .map(criterion -> left.getOutputSymbols().contains(criterion.getLeft()) ? criterion : criterion.flip())
+                    .collect(toImmutableList());
+        }
+
+        private static PlanNode getJoinSource(PlanNodeIdAllocator idAllocator, List<PlanNode> nodes, Expression filter, List<Symbol> outputSymbols)
+        {
+            PlanNode planNode;
+            if (nodes.size() == 1) {
+                planNode = getOnlyElement(nodes);
+                if (!(BooleanLiteral.TRUE_LITERAL).equals(filter)) {
+                    return new FilterNode(idAllocator.getNextId(), planNode, filter);
+                }
+                return planNode;
+            }
+            return new JoinGraphNode(idAllocator.getNextId(), nodes, filter, outputSymbols);
```

Comment:
why not immediately do recursion here and expand the tree instead of returning `JoinGraphNode`? This way you wouldn't need to make `JoinGraphNode` a `PlanNode` and it wouldn't have to be public. 

Also this would be more performant since no additional `JoinGraphNodes` would need to be created and no additional `replaceChildren` would need to be called. I prefer recursion (and more compact algorithm) instead of two phase tree expansion with https://github.com/Teradata/presto/pull/559/commits/62adaddf85df9e5d5c71ce0426681131afc6c2e2#diff-5cdb22a9680530f837d520839d7eaf99R256